push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/MetricsCollector.java

```diff
@@ -65,18 +74,52 @@
      * Additionally when a variant was completed and there are no more variants to be completed for this variant,
      * the {@link PushMessageCompletedEvent} CDI event is fired.
      *
-     * @param variantMetricInformation the variant metrics info object
+     * @param event {@link TriggerMetricCollectionEvent} event dequeued from JMS
+     * @throws JMSException when JMS provider fails to dequeue messages that {@link MetricsCollector} pulls
      */
-    public void collectMetrics(@Observes @Dequeue VariantMetricInformation variantMetricInformation) {
-        PushMessageInformation pushMessageInformation = metricsService.lock(variantMetricInformation.getPushMessageInformation());
-        pushMessageInformation.setTotalReceivers(pushMessageInformation.getTotalReceivers() + variantMetricInformation.getReceivers());
+    public void collectMetrics(@Observes @Dequeue TriggerMetricCollectionEvent event) throws JMSException {
+        final String pushMessageInformationId = event.getPushMessageInformationId();
+        final PushMessageInformation pushMessageInformation = metricsService.getPushMessageInformation(pushMessageInformationId);
+        metricsService.lock(pushMessageInformation);
+
+        receiveVariantMetricsRemainingInQueues(pushMessageInformation);
+
+        for (VariantMetricInformation variantMetricInformation : pushMessageInformation.getVariantInformations()) {
+            if (areAllBatchesLoaded(variantMetricInformation)) {
+                pushMessageInformation.setServedVariants(1 + pushMessageInformation.getServedVariants());
+                logger.trace(String.format("All batches for variant %s were processed", variantMetricInformation.getVariantID()));
+                variantCompleted.fire(new VariantCompletedEvent(pushMessageInformation.getId(), variantMetricInformation.getVariantID()));
+            }
+        }
+
+        if (areAllVariantsServed(pushMessageInformation)) {
+            event.markAllVariantsProcessed();
+            logger.trace(String.format("All variants for application %s were processed", pushMessageInformation.getId()));
+            pushMessageCompleted.fire(new PushMessageCompletedEvent(pushMessageInformation.getId()));
+        }
 
-        // AGPUSH-1585:
-        // using a combined key of variant ID and PushMessageInformation ID, to not limit different push requests on the queue just to the variant
-        // TODO: improve name and/or implementation if this.
-        final String variantPushMessageID = variantMetricInformation.getVariantID() + ":" + pushMessageInformation.getId();
+        metricsService.updatePushMessageInformation(pushMessageInformation);
+    }
 
-        int loadedBatches = countLoadedBatches(variantPushMessageID);
+    private void receiveVariantMetricsRemainingInQueues(PushMessageInformation pushMessageInformation) throws JMSException {
+        while (true) {
```

Comment:
I'd hate to change this.. it's working fine on MASTER.

However,I am fine in doing a fix for that on master. with a new AGPUSH JIRA